Add ngsim documentation (#1845)
huawei-noah/SMARTS: docs/ecosystem/ngsim.rst

> .. _ngsim:

Reviewer: Document isn't included in any toctree. We may include it under the ecosystem toctree in index.rst .

Author: Done.

NGSIM
=====

**NGSIM** is the Next Generation Simulation dataset, a free use dataset created
 by the US Department of Transportation (DOT) and described
`here <https://ops.fhwa.dot.gov/trafficanalysistools/ngsim.htm>`_.

NGSIM traffic history scenarios
-------------------------------

These scenarios are located in the `smarts repository <https://github.com/huawei-noah/SMARTS/tree/master/scenarios/NGSIM>`_.

To use these scenarios, you must first download the dataset in an appropriate format.
SMARTS assumes that the data will be in tab-separated text files.
Its import scripts were tested on and work with the data downloaded from `here <http://bit.ly/PPUU-data>`_,
which is the same as was used by this `PPUU project <https://github.com/Atcold/pytorch-PPUU>`_.

After the dataset has been downloaded, you will need to update each ``scenario.py``

> in the ngsim scenario directories (e.g., :scenarios:`NGSIM/i80/scenario.py`, :scenarios:`NGSIM/us101/scenario.py`, and/or :scenarios:`NGSIM/peach/scneario.py`) to point the `input_path` field

Reviewer: ```suggestion
in the ngsim scenario directories (e.g., :scenarios:`NGSIM/i80/scenario.py`, :scenarios:`NGSIM/us101/scenario.py`, and/or :scenarios:`NGSIM/peachtree/scenario.py`) to point the `input_path` field
```

to the dataset location on your filesystem.

> Once that is done, you should be able to build your scenarios in the normal way, for example:
> ```bash
> scl scenario build-all --clean scenarios/NGSIM
> ```

Reviewer: ```suggestion
Once that is done, you should be able to build your scenarios in the normal way, for example: ``scl scenario build-all --clean scenarios/NGSIM``
```

For each traffic history dataset specified in your ``scenario.py``, 
this will create a corresponding ``.shf`` file that SMARTS will use
whenever this traffic will be added to a simulation.

Note that the SUMO maps created for NGSIM were made by hand by the open-source SMARTS team.
Although they attempt to align with the positions in the traffic dataset,
their level of exactness may not be enough for some model-training situations,
so you may want or need to refine them with SUMO's `netedit tool <https://sumo.dlr.de/docs/Netedit/index.html>`_.

> An example of how traffic history might be saved as observations can be found in :py:module:`smarts.dataset.traffic_histories_to_observations`. 

Author: Do you happen to have the documentation that mentions the details of this syntax? I have found that it is possible from some sources but not the list of reference types.

> An example of how traffic history might be saved as observations can be found in :py:module:`smarts.dataset.traffic_histories_to_observations`. 

Reviewer: Cross-referencing Python objects
https://www.sphinx-doc.org/en/master/usage/restructuredtext/domains.html#cross-referencing-python-objects

                assert match is not None
                vehicle_id = match.group(1)
                if vehicle_id not in vehicle_ids:
                    vehicle_ids.append(vehicle_id)

        for id_ in vehicle_ids:
            with open(scenario_path / f"{id}.pkl", "rb") as f:
                vehicle_data = pickle.load(f)

            image_names = list()
            for filename in os.listdir(scenario_path):
                if filename.endswith(f"{id_}.png"):
                    image_names.append(filename)
            image_names.sort()

            for i in range(len(image_names)):
                with Image.open(scenario_path / image_names[i], "r") as image:
                    image.seek(0)
                    bev = np.moveaxis(np.asarray(image, dtype=np.uint8), -1, 0)
                
                sim_time = image_names[i].split("_")[0]
                current_obs: Observation = vehicle_data[float(sim_time)]
                obs.append((current_obs, bev))

        for o, rgb in obs:
            ...

Alternatively, an approach like :examples:`traffic_histories_vehicle_replacement.py` can be used to operate directly with the scenarios.

> Samples
> ------

Reviewer: Title underline is short.
```suggestion
Samples
-------
```
